src/main/java/org/apache/maven/plugins/dependency/tree/verbose/VerboseGraphSerializer.java

```diff
@@ -0,0 +1,323 @@
+        {
+            scope = node.getDependency().getScope();
+        }
+
+        String coords = artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getExtension() + ":"
+                + artifact.getVersion();
+
+        if ( scope != null && !scope.isEmpty() )
+        {
+            coords = coords.concat( ":" + scope );
+        }
+        return coords;
+    }
+
+    private static String getVersionlessScopelessCoordinate( DependencyNode node )
+    {
+        Artifact artifact = node.getArtifact();
+        // scope not included because we check for scope conflicts separately
+        return artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getExtension();
+    }
+
+    private static boolean isDuplicateDependencyCoordinate( DependencyNode node, Set<String> coordinateStrings )
+    {
+        return coordinateStrings.contains( getDependencyCoordinate( node ) );
+    }
+
+    private static String versionConflict( DependencyNode node, Map<String, String> coordinateVersionMap )
+    {
+        if ( coordinateVersionMap.containsKey( getVersionlessScopelessCoordinate( node ) ) )
+        {
+            return coordinateVersionMap.get( getVersionlessScopelessCoordinate( node ) );
+        }
+        return null;
+    }
+
+    private static String scopeConflict( DependencyNode node, Set<String> coordinateStrings )
+    {
+        Artifact artifact = node.getArtifact();
+        List<String> scopes = Arrays.asList( "compile", "provided", "runtime", "test", "system" );
+
+        for ( String scope : scopes )
+        {
+            String version;
+            if ( artifact.getProperties().containsKey( "version" ) )
+            {
+                version = artifact.getProperties().get( "version" );
+            }
+            else
+            {
+                version = artifact.getVersion();
+            }
+
+            String coordinate = artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getExtension()
+                    + ":" + version + ":" + scope;
+            if ( coordinateStrings.contains( coordinate ) )
+            {
+                return scope;
+            }
+        }
+        return null;
+    }
+
+    private Map<DependencyNode, String> getNodeConflictMessagesBfs( DependencyNode root, Set<String> coordinateStrings
+            , Map<String, String> coordinateVersionMap )
+    {
+        Map<DependencyNode, String> nodeErrors = new HashMap<>();
+        Set<DependencyNode> visitedNodes = new HashSet<>( 512 );
+        Queue<DependencyNode> queue = new LinkedList<>();
+        visitedNodes.add( root );
+        queue.add( root );
+
+        while ( !queue.isEmpty() )
+        {
+            DependencyNode node = queue.poll();
+
+            if ( node == null || node.getArtifact() == null )
+            {
+                // Should never reach hit this condition with a proper graph sent in
+                nodeErrors.put( node, "Null Artifact Node" );
```

Comment:
Is this awful enough to warrant just throwing?

Reply:
My rationale for not throwing an exception is since this is the verbose option if a user is running this then something has probably gone wrong with their dependency resolution and we should show them if the graph has an invalid node somewhere so they can look into it.